File: heap/1962_remove_stones_to_minimize_the_total.py
class Solution:
    # O(klog(n)) time | O(n) space - where n is piles.length
    def minStoneSum(self, piles: List[int], k: int) -> int:
        maxHeap = []
        for stones in piles:
            heapq.heappush(maxHeap, -stones)
        
        for _ in range(k):
            stones = -heapq.heappop(maxHeap)
            stones -= stones // 2
            heapq.heappush(maxHeap, -stones)
        
        return -sum(maxHeap)

File: heap/1962_remove_stones_to_minimize_the_total.java
class Solution {
    // O(klog(n)) time | O(n) space - where n is piles.length
    public int minStoneSum(int[] piles, int k) {
        PriorityQueue<Integer> heap = new PriorityQueue<>((a, b) -> b - a);
        for (int num: piles) {
            heap.add(num);
        }

        for (int i = 0; i < k; i++) {
            int stones = heap.remove();
            stones -= stones / 2;
            heap.add(stones);
        }
        int totalStones = 0;
        for (int stone: heap) {
            totalStones += stone;
        }
        return totalStones;
    }
}
